# Agglomerative Clustering



 این الگوریتم با هدف تحلیل شخصیتی مشتری وتقسیم بندی انها برای بازاریابی بهتر برای کارکنان و سرمایه داران طراحی شده است. داده شامل اطلاعاتی از قبیل  درآمد مشتری , تعداد فرزندان مشتری,سال تولد و... را شامل میشود.

This algorithm is designed to analyze customer personality and segment them for better marketing for employees and investors. The data includes information such as customer income, number of children, year of birth, and so on.



 توجه توضیح یسری از سلول ها اگر وجود نداره در فایل

music.ipynb
 هست


 Note: Some of the explanations or cells may be missing here, but they are available in the file `music.ipynb`.

In [1]:
import pandas as pd
import numpy as np
import scipy
import matplotlib.cm as cm
import matplotlib.pyplot as plt
import  matplotlib.pylab as pylab

دیتا از:https://www.kaggle.com/code/karnikakapoor/customer-segmentation-clustering
~~~ Data


ستونDt_Customer به صورت دستی حذف شد.

The column Dt_Customer was deleted manually.

In [2]:
x=pd.read_csv('marketing_campaign.csv')
x

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,58,635,88,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,38,11,1,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,26,426,49,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,26,11,4,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,94,173,43,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1967,Graduation,Married,61223.0,0,1,46,709,43,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1946,PhD,Together,64014.0,2,1,56,406,0,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1981,Graduation,Divorced,56981.0,0,0,91,908,48,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1956,Master,Together,69245.0,0,1,8,428,30,...,3,0,0,0,0,0,0,3,11,0


In [3]:
x.isnull().sum()

ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64

In [4]:
x=x.dropna()
x

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,58,635,88,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,38,11,1,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,26,426,49,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,26,11,4,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,94,173,43,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1967,Graduation,Married,61223.0,0,1,46,709,43,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1946,PhD,Together,64014.0,2,1,56,406,0,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1981,Graduation,Divorced,56981.0,0,0,91,908,48,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1956,Master,Together,69245.0,0,1,8,428,30,...,3,0,0,0,0,0,0,3,11,0


In [5]:
x=x.drop('ID',axis=1)

x

,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Recency,MntWines,MntFruits,MntMeatProducts,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,1957,Graduation,Single,58138.0,0,0,58,635,88,546,...,7,0,0,0,0,0,0,3,11,1
1,1954,Graduation,Single,46344.0,1,1,38,11,1,6,...,5,0,0,0,0,0,0,3,11,0
2,1965,Graduation,Together,71613.0,0,0,26,426,49,127,...,4,0,0,0,0,0,0,3,11,0
3,1984,Graduation,Together,26646.0,1,0,26,11,4,20,...,6,0,0,0,0,0,0,3,11,0
4,1981,PhD,Married,58293.0,1,0,94,173,43,118,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,1967,Graduation,Married,61223.0,0,1,46,709,43,182,...,5,0,0,0,0,0,0,3,11,0
2236,1946,PhD,Together,64014.0,2,1,56,406,0,30,...,7,0,0,0,1,0,0,3,11,0
2237,1981,Graduation,Divorced,56981.0,0,0,91,908,48,217,...,6,0,1,0,0,0,0,3,11,0
2238,1956,Master,Together,69245.0,0,1,8,428,30,214,...,3,0,0,0,0,0,0,3,11,0


In [6]:
# x=np.asanyarray(x)

In [7]:
from sklearn.preprocessing import LabelEncoder

lb=LabelEncoder()
x['Education']=lb.fit_transform(x['Education'])


lb2=LabelEncoder()
x['Marital_Status']=lb2.fit_transform(x['Marital_Status'])

In [8]:
from sklearn.preprocessing import MinMaxScaler
sc=MinMaxScaler()
x=sc.fit_transform(x)
x

array([[0.62135922, 0.5       , 0.57142857, ..., 0.        , 0.        ,
        1.        ],
       [0.59223301, 0.5       , 0.57142857, ..., 0.        , 0.        ,
        0.        ],
       [0.69902913, 0.5       , 0.71428571, ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.85436893, 0.5       , 0.28571429, ..., 0.        , 0.        ,
        0.        ],
       [0.61165049, 0.75      , 0.71428571, ..., 0.        , 0.        ,
        0.        ],
       [0.59223301, 1.        , 0.42857143, ..., 0.        , 0.        ,
        1.        ]], shape=(2216, 27))

In [9]:
from sklearn.decomposition import PCA
pca=PCA(n_components=2)
x=pca.fit_transform(x)
x

array([[ 0.74887008,  0.30531189],
       [-0.53869989,  0.11122599],
       [ 0.23751754, -0.30500341],
       ...,
       [ 0.47438195, -0.4563821 ],
       [ 0.18941389, -0.30715032],
       [-0.05427714,  0.78489996]], shape=(2216, 2))

In [10]:
from sklearn.metrics.pairwise import euclidean_distances
dist_matrix = euclidean_distances(x,x)
print(dist_matrix)

[[0.         1.30211588 0.79621994 ... 0.8096428  0.82951865 0.93544117]
 [1.30211588 0.         0.88077263 ... 1.16125525 0.83975497 0.82976022]
 [0.79621994 0.88077263 0.         ... 0.28110542 0.04815154 1.12828786]
 ...
 [0.8096428  1.16125525 0.28110542 ... 0.         0.32167829 1.3491707 ]
 [0.82951865 0.83975497 0.04815154 ... 0.32167829 0.         1.11890979]
 [0.93544117 0.82976022 1.12828786 ... 1.3491707  1.11890979 0.        ]]


In [11]:
from sklearn.cluster import AgglomerativeClustering
cluster_n=4
model=AgglomerativeClustering(linkage='average',n_clusters=cluster_n).fit(x)
print(model.labels_)

[1 0 0 ... 0 0 2]


In [12]:
label_ll=model.labels_
label_ll

array([1, 0, 0, ..., 0, 0, 2], shape=(2216,))



در این قسمت خوشه بندی نمره خوبی دریافت نکرده است وضعیف عمل کرده است.

"In this clustering section, it has not received a good score and has performed poorly.

In [13]:
from sklearn.metrics import silhouette_score
silhouette_score(x, label_ll)


0.43545029699535504


در اینجا هم مدل عملکرد خوبی نداشته و نمره پایینی دریافت کرده.

here too, the model has not performed well and has received a low score.

In [14]:
from sklearn.metrics import davies_bouldin_score

davies_bouldin_score(x, label_ll)

0.6696462734315035


برخلاف دو معیار ارزیابی فوق، این مدل در جداسازی خوشه‌ها عملکرد بسیار خوبی داشته و امتیاز خوبی دریافت کرده است.

Unlike the two evaluation metrics above, the model has performed excellently in keeping clusters apart and has received a good score.

In [15]:
from sklearn.metrics import calinski_harabasz_score

calinski_harabasz_score(x, label_ll)

763.6971050854825



بطور کلی , مدل عالی عمل نکرده است و تنها درجداسازی خوشه ها خوب عمل کرده.

overall, the model has not performed excellently and has only done well in separating the clusters.

در اینجا میبینم کلاسورینگ چجوری خوشه بندی کرده و هر گروه دقیقا چند نفر رو داره.
گروه 0 از همه بیشتر هستش حالا بیاید چک کنیم ببینم این گروه ها چه کسایی هستن و چرا برا بیزنس ما خوبن


Here, I can see how the clustering algorithm has grouped the data and exactly how many people are in each group

Group 0 is the largest group. Now, let's check who these groups are and why they are good for our business


In [16]:
unique,counts=np.unique(model.labels_, return_counts=True)
print(dict(zip(unique,counts)))

{np.int64(0): np.int64(1868), np.int64(1): np.int64(218), np.int64(2): np.int64(118), np.int64(3): np.int64(12)}


In [17]:
df=pd.read_csv('marketing_campaign.csv')
df=df.dropna()
df=df.drop('ID',axis=1)

df['Cluster']=model.labels_

cluster_summary=df.groupby('Cluster').mean(numeric_only=True)
print(cluster_summary.T)

Cluster                         0             1             2             3
Year_Birth            1968.716274   1968.330275   1971.593220   1966.666667
Income               50560.124197  74247.646789  35025.796610  84547.166667
Kidhome                  0.462527      0.045872      0.889831      0.000000
Teenhome                 0.543897      0.243119      0.423729      0.083333
Recency                 51.475910     38.720183     28.474576     54.500000
MntWines               264.589400    730.032110     96.127119    945.000000
MntFruits               23.817452     54.211009      7.728814     98.666667
MntMeatProducts        141.501071    427.472477     47.855932    575.250000
MntFishProducts         34.374197     74.995413      8.627119    152.250000
MntSweetProducts        24.422377     54.821101      7.779661    117.166667
MntGoldProds            40.489829     76.440367     33.372881     99.166667
NumDealsPurchases        2.330835      1.756881      3.389831      1.000000
NumWebPurcha

In [18]:
total_income=df.groupby(model.labels_)['Income'].sum()

for cluster, income in total_income.items():
    print(f"Cluster {cluster} : {income:,.2f}")

Cluster 0 : 94,446,312.00
Cluster 1 : 16,185,987.00
Cluster 2 : 4,133,044.00
Cluster 3 : 1,014,566.00


In [24]:
df["all_mtn"]=df["MntWines"]+df["MntFruits"]+df["MntMeatProducts"]+df["MntFishProducts"]+df["MntSweetProducts"]+df["MntGoldProds"]
df["child_numbers"]=df["Kidhome"]+df["Teenhome"]

a_mtn=df.groupby("Cluster")['all_mtn'].mean()
a_child_num=df.groupby("Cluster")['child_numbers'].mean()

for i in  a_mtn.index:
    print(f"mean:{i} cluster mtn:{a_mtn[i]},child_mean:{a_child_num[i]}")

mean:0 cluster mtn:529.1943254817987,child_mean:1.006423982869379
mean:1 cluster mtn:1417.9724770642201,child_mean:0.2889908256880734
mean:2 cluster mtn:201.4915254237288,child_mean:1.3135593220338984
mean:3 cluster mtn:1987.5,child_mean:0.08333333333333333


Group 0: Income and expenditure: compared to the other groups, they take second rank in expenditure and income, and their number is the highest of all. They did not participate in the survey regarding whether they cooperate or not.

Group 1: relatively good income and very good expenditure, with a relatively high number (218 people). They fully participated in the survey and said they cooperate.

Group 2: lower income compared to the other groups, and also had the lowest expenditure. They did participate in the survey either.

Group 3: income lower than the other groups, but expenditure is drastically higher. They did not participate in the survey.


Conclusion and my personal opinion: if we want to work on a combined basis for both expenditure and financial support, as well as insurance and investment, and want both to be present, Group 0, given the high number of people and their income and expenditure, is very good both for insurance and for financial support and investment. If we can satisfy them and find out the reason why they did not participate in the survey, we can have an upward trend with them. But if our goal is only expenditure and sales, Groups 1, 3, and 2 are the best and most reasonable of all, and Group 3 should also be satisfied and talked to, to see what the reason for their non-participation is. Groups 1, 3, and 2 cannot be counted on much for insurance and long-term investment; they are better for selling goods, if of course that is our goal.

 گروه 0درامد وخرج: نسبت به بقیه گروه ها رتبه دوم در خرج و درامد رو میگیرن و تعدادشون از همه بیشتره. در نظرسنجی شرکت نکردن که همکاری میکنن یا نه

. گروه 1: درامدنسبتا خوب و خرج بسیار خوب با تعداد نسبتا بالا(218 نفر) در نظرسنجی هم کاملا شرکت کردن و گفتن همکاری میکنن.

گروه 2: درامد پایین تر نسبت به گروه های دیگ و همچین کمترین خرج رو داشتن در نظرسنجی هم شرکت کردن.

 گروه3: درامد کمتر از بقیه گروه ها ولی خرج به شدت بالاتر در نظرسنجی شرکت نکردن.


نتیجه گیری و دیدگاه شخصی من: اگر برای خرج و ساپورت مالی  و هم بیمه و سرمایه گذاری  بخوایم ترکیبی کار کنیم و هردو باشه گروه 0 با توجه به تعداد بالا  افراد و درامد و خرجشون هم برای بیمه و هم برای ساپورت مالی و  سرمایه گذاری بسیار خوب هستن اگر بتونیم راضیشون کنیم و بفهمیم دلیل شرکت نکردشون در نظرسنجی چیه میتونیم روند صعودی داشته باشیم باهاشون . اما اگر هدف ما  فقط خرج و فروش باشه گروه1 و 3و2 از همه بهتر و معقول تر هست و باید گروه 3 هم راضی بشن و باهاشون حرف زده بشه ببینم دلیل شرکت نکردنشون چیه.  روی گروه 1 و 3و2 زیاد نمیشه رو بیمه و سرمایه گذاری بلند مدت حساب کرد بیشتر برای فروش اجناس خوبن اگر البته هدف ما این باشه.